lib/net/tti/messages/function_calls/authentication.rb

```diff
@@ -3,29 +3,40 @@
 module Net
   module TTI
     class Authentication < FunctionCall
-      LOGON_MODE_PRE_AUTH   = 0x00000001
-      LOGON_MODE_AUTH       = 0x00000101
-
-      def self.create_pre_auth_request(target_architecture)
-        case target_architecture
-        when :x86
-          return PreAuthenticationX86.new
-        when :x64
-          return PreAuthenticationX64.new
-        else
-          raise Net::TTI::Exceptions::UnsupportedPlatform.new(target_architecture)
-        end
-      end
-
-      def self.create_auth_request(target_architecture)
-        case target_architecture
-        when :x86
-          return AuthenticationX86.new
-        when :x64
-          return AuthenticationX64.new
-        else
-          raise Net::TTI::Exceptions::UnsupportedPlatform.new(target_architecture)
-        end
+      LOGON_MODE_PRE_AUTH   = 0x01
+      LOGON_MODE_AUTH       = 0x0101
+      uint8     :unknown1,              :initial_value => 0x01
+      uint8     :username_length_size,  :initial_value => 0x01
+      uint8     :username_length,       :value => lambda { username.length }
+      uint8     :logon_mode_size,       :initial_value => lambda { _logon_mode == Authentication::LOGON_MODE_PRE_AUTH ? 0x01: 0x02}
+      choice    :logon_mode,            :selection => :_logon_mode do
+        uint8    LOGON_MODE_PRE_AUTH,   :initial_value => lambda { _logon_mode }
+        uint16le LOGON_MODE_AUTH,       :initial_value => lambda { _logon_mode }
+      end
+      uint8     :unused1,               :initial_value => 0x01
+      uint8     :parameters_count_size, :initial_value => 0x01
+      uint8     :parameters_count,      :value => lambda {parameters.count}
+      uint8     :unused2,               :initial_value => 0x01
+      uint8     :unused3,               :initial_value => 0x01      
+      string    :username
+      array     :parameters,   :type => :key_value_pair, :read_until => lambda {index == parameters_count - 1}
+
```

Comment:
I tried to mimic JDBC driver behavior here, it results in different messages exchanged with server compared to OCI applications (e.g. sqlplus) indeed but my hope is that JDBC uses "universal" format that all servers understand.